Display stream and topic name as title on action sheet.

hashirsarwar opened the pull request:
This PR is followed by #3707.
Fixes #3631.

When a user clicks on the topic header, the action sheet displays a title with stream and topic name in the format `#stream > topic`.

Here's the screenshot:

![screenshot-2019-12-10_10 02 21 828](https://user-images.githubusercontent.com/45683359/70497932-9928e700-1b36-11ea-8948-af42ed3bcecc.png)

hashirsarwar commented:
@gnprice kindly review it.
And please note, I chose `v2.0.1` over `v3.4.0` because:

- `v2.0.1` uses the same old dependencies, while `v3.4.0` adds some new dependencies to the project as you can see [here](https://github.com/zulip/zulip-mobile/pull/3707/files/d13ddcfa4e37fb341129d3c849f94852bf214d60#diff-8ee2343978836a779dc9f8d6b794c3b2R842-R2815).

- According to the [documentation](https://www.npmjs.com/package/@expo/react-native-action-sheet), I felt that `v3.4.0` has slightly different way to setup an action sheet in a project; whereas, `v2.0.1` and `v1.1.0` are exactly same. So, `v3.4.0` tend to have some compatibility issues with the project.

- Right now, we are only concerned about the `title` option for action sheet and it is being provided by `v2.0.1` without any issue.

jainkuniya commented:
Can you please also add a screenshot, how it looks when topic is long,
maybe for iOS too.

On Mon, Dec 9, 2019, 3:24 AM Hashir Sarwar <notifications@github.com> wrote:

> @gnprice <https://github.com/gnprice> kindly review it.
> And please note, I chose v2.0.1 over v3.4.0 because:
>
>    -
>
>    v2.0.1 uses the same old dependencies, while v3.4.0 adds some new
>    dependencies to the project as you can see here
>    <https://github.com/zulip/zulip-mobile/pull/3707/files/d13ddcfa4e37fb341129d3c849f94852bf214d60#diff-8ee2343978836a779dc9f8d6b794c3b2R842-R2815>
>    .
>    -
>
>    According to the documentation
>    <https://www.npmjs.com/package/@expo/react-native-action-sheet>, I
>    felt that v3.4.0 has slightly different way to setup an action sheet
>    in a project; whereas, v2.0.1 and v1.1.0 are exactly same. So, v3.4.0
>    tend to have some compatibility issues with the project.
>    -
>
>    Right now, we are only concerned about the title option for action
>    sheet and it is being provided by v2.0.1 without any issue.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip-mobile/pull/3714?email_source=notifications&email_token=AENHKSI75T4MVD243GSW6RLQXVUIJA5CNFSM4JXSCXHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGHLBUQ#issuecomment-562999506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AENHKSPQS7SGA24OQSRYNKDQXVUIJANCNFSM4JXSCXHA>
> .
>

gnprice commented:
Copying from a thread on the previous PR #3707:
> The dependencies are fine. Let's use the latest version.
>
> I'm still curious, though: why did they increase the major version number? That generally means there's some change that they expected to break some existing code that uses the library. It'd be good to know what that change was, so we can check if it affects us and there's something we need to do to adjust for it.

The difference in setup instructions looks like just a tweak to how the instructions are written, not an actual change in the API, so I don't think that's it. (In any case it would still leave unanswered the question of how v2 differs from v1.)

hashirsarwar commented:
@gnprice the library has not provided release notes for the older verions. 
But if we [compare the v1.1.2 with master](https://github.com/expo/react-native-action-sheet/compare/v1.1.2...master), we'll find out the commits made after the release of v1.1.2.

Now the 29 commits starting from the top until [afb0fef - Updated version to 2.0.0](https://github.com/expo/react-native-action-sheet/commit/afb0fefd7b56bc1c6384bd6e14cdb8e9dbf59273) are the commits between v1 and v2. Some of the important changes that I felt are:
- Adds Android support for an optional title prop, similar to native iOS title. [9fee7dc](https://github.com/expo/react-native-action-sheet/commit/9fee7dc244569ded220c8d28286118a5f6514a13)
- Adds support for message prop in Android. [3815016](https://github.com/expo/react-native-action-sheet/commit/38150163c1fa9af77064f40f92e7b1d95363c805)
- Allows Android textStyle to be declared at the Provider. [dfe7242](https://github.com/expo/react-native-action-sheet/commit/dfe724247ceec13fa34a712dc5a661157b3b2fe9)
- Resolved merge conflict. [150cf5f](https://github.com/expo/react-native-action-sheet/commit/150cf5fcca08df04dcb9e8d26b489af79c04d9e7)
- Added new Android props: title, message, titleTextStyle, messageTextStyle, showSeparators (defaults to false). [b2862ba](https://github.com/expo/react-native-action-sheet/commit/b2862ba08ca930d1b986380e752d71f156fad87d)
- Fixed an existing crash where icons actually crashed due if you didn’t provide the textStyle prop. [b2862ba](https://github.com/expo/react-native-action-sheet/commit/b2862ba08ca930d1b986380e752d71f156fad87d)
- Fixed a crash on iOS if you passed a null title or message. [b2862ba](https://github.com/expo/react-native-action-sheet/commit/b2862ba08ca930d1b986380e752d71f156fad87d)
- Added tintColor to Android to match iOS option. [acfd89c](https://github.com/expo/react-native-action-sheet/commit/acfd89c465df637a67578b612ff8397bd6eea16c)
- Use shorthand for null to set title and message to undefined. [7fe0aff](https://github.com/expo/react-native-action-sheet/commit/7fe0aff0fe83ce093bb25884ee5c0b4d61fce032)
- Added new Android-only prop: tintIcons. [3c95b8a](https://github.com/expo/react-native-action-sheet/commit/3c95b8a86d746c2a876dc9910c9848e244a8f854)
- Use !! with && to render instead of ternary defaulting to null. [59e34c8](https://github.com/expo/react-native-action-sheet/commit/59e34c83b457910bf2f4ea47bd0db80649c3f83b)

hashirsarwar commented:
@gnprice and please note, when a new topic is being added, user can't enter 54+ words in the web app which is pretty normal. But in case of mobile app, there is no limit. Though, if user enters more than 54 words the topic string is automatically truncated (replaced with ... after 54 words) while sending it to the server. We can also add a limit for mobile app just like the web app.

hashirsarwar commented:
> Can you please also add a screenshot, how it looks when topic is long, maybe for iOS too.
> […](#)
> On Mon, Dec 9, 2019, 3:24 AM Hashir Sarwar ***@***.***> wrote: @gnprice <https://github.com/gnprice> kindly review it. And please note, I chose v2.0.1 over v3.4.0 because: - v2.0.1 uses the same old dependencies, while v3.4.0 adds some new dependencies to the project as you can see here <https://github.com/zulip/zulip-mobile/pull/3707/files/d13ddcfa4e37fb341129d3c849f94852bf214d60#diff-8ee2343978836a779dc9f8d6b794c3b2R842-R2815> . - According to the documentation ***@***.***/react-native-action-sheet>, I felt that v3.4.0 has slightly different way to setup an action sheet in a project; whereas, v2.0.1 and v1.1.0 are exactly same. So, v3.4.0 tend to have some compatibility issues with the project. - Right now, we are only concerned about the title option for action sheet and it is being provided by v2.0.1 without any issue. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <#3714?email_source=notifications&email_token=AENHKSI75T4MVD243GSW6RLQXVUIJA5CNFSM4JXSCXHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGHLBUQ#issuecomment-562999506>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AENHKSPQS7SGA24OQSRYNKDQXVUIJANCNFSM4JXSCXHA> .

Updated the screenshot.

gnprice commented:
> @gnprice the library has not provided release notes for the older verions.
> But if we [compare the v1.1.2 with master](https://github.com/expo/react-native-action-sheet/compare/v1.1.2...master), we'll find out the commits made after the release of v1.1.2.
> 
> [...]

Thanks for going through this comparison! That's very informative.

None of those changes sound like they'd be incompatible with code that was written for previous versions, and so would be a reason to bump the major version number. :man_shrugging: I do see in that commit log at least a hint of why they might have thought it was a good idea to go from v2 to v3: commit https://github.com/expo/react-native-action-sheet/commit/16d1ef0b41e0f59e2ac4173e8db4565382afbec8 is titled "Publish 3.0.0 with TypeScript!"

Hmm, and a bit before that, they had https://github.com/expo/react-native-action-sheet/commit/40f945a035c83745934f3be6db41d8c1e4db747d ":boom: (src) convert flow to ts". Yeah, that does sound like it could potentially affect how our app interacts with it, at least in type-checking -- they used to have Flow source code (same as us), and now they have TS source code.

The code in this PR looks good to me, and so does the upgrade commit AFAICT, with just one comment:
* In the commit message, adding a line like this would be helpful:
  > This library doesn't publish release notes, but there don't appear to have been any breaking changes between these versions. For details, see #3714.

  That states the key conclusion, and also helps a future reader find your detailed comment here if they want to better understand what we were thinking :slightly_smiling_face: 

Aside from that comment, there are just two things I'd want to do before merging this:
* Briefly play with it in an emulator and check the UI seems good.
* Check into how the new TS version interacts with our type-checking, and in particular see if there's anything we should do to our libdefs so we keep reasonable type-checking for our use of this library.

I am just about to go on vacation, as it happens, so won't be able to do those soon. I'll be back in a couple of weeks, so I'll get to it as I work through my queue after that; or possibly @ray-kraesig will take care of it before then.

Thanks again for the contributions!

hashirsarwar commented:
@gnprice thank you for the review. I have made the required changes in the commit message. Have a nice vacation! :slightly_smiling_face:

ray-kraesig commented:
> I am just about to go on vacation, as it happens, so won't be able to do those soon. I'll be back in a couple of weeks, so I'll get to it as I work through my queue after that; or possibly @ray-kraesig will take care of it before then.

:wave:

> * Briefly play with it in an emulator and check the UI seems good.

*Mostly* :+1:. I don't think the format suggestion from https://github.com/zulip/zulip-mobile/pull/3707#pullrequestreview-328451150 has turned out well – I'd prefer to drop the initial `#${params.message.display_recipient} >` part, for approximately two-and-a-half reasons:

* The current stream name is still visible at the top of the screen, so it's redundant.
* The display format is technically ambiguous if either the stream or the topic name contains `" > "` as a substring.
* The type of `Message.display_recipient` is a mess. This assumes it's `string` – which is in fact correct, here, but I don't look forward to trying to explain that to Flow once we start cleaning that type up.

I suppose just `params.message.subject` might be confusing if the topic were named something deliberately ambiguous, but I think the additional context that the user has – that of having just long-pressed the topic header, which *does* already have at least a partial display of the topic name – should be enough to disambiguate. At most, something like `` `topic: ${params.message.subject}` `` should suffice.

> * Check into how the new TS version interacts with our type-checking, and in particular see if there's anything we should do to our libdefs so we keep reasonable type-checking for our use of this library.

Objection: statement assumes facts not in evidence. We don't actually *have* reasonable type-checking for this library as it stands – v1.1.0 has a `.flowconfig`, but doesn't seem to export any Flow types or provide any typing information. :disappointed:

On the other hand, that does mean that the upgrade won't make things any worse on that front. I'm not inclined to say type definitions are needed in this PR.

hashirsarwar commented:
> The display format is technically ambiguous if either the stream or the topic name contains " > " as a substring.

:+1: 

> At most, something like `topic: ${params.message.subject}` should suffice.

@ray-kraesig I think it would be better if we somehow distinguish `topic:` from the topic's name. We can use bold text or change the color to dark black for `topic:`.

hashirsarwar commented:
@ray-kraesig please have a look.

![831a913b-c976-43cd-884d-6a5576fc3e1b](https://user-images.githubusercontent.com/45683359/71449359-81737480-276c-11ea-83c1-143c7a27b68c.jpeg)

ray-kraesig commented:
> > At most, something like `topic: ${params.message.subject}` should suffice.
> 
> @ray-kraesig I think it would be better if we somehow distinguish `topic:` from `params.message.subject`. We could use bold text or change the color to dark black for `topic:` but it is not possible to alter the text configuration in this library.

For reference, this is mostly because the library is built on top of the native iOS Action Sheet object, which [doesn't support rich text](https://developer.apple.com/documentation/uikit/uialertcontroller/1620103-title).

>  Anyways, I'll post a screenshot for you to check whether your suggested change looks good.

Reasonable! I'm of the opinion that `""` > `"topic: "` > `"Topic: "` – but it's not a very *strong* opinion, and I'd be happy with any of the three. If you use one of the latter two, be sure to wrap it in `_(...)`, for internationalization's sake.

On the other hand, @gnprice should be back from vacation fairly soon, and I'm sure he'll want to weigh in on this.

gnprice commented:
> * The current stream name is still visible at the top of the screen, so it's redundant.

It might be, but there's also the "All messages" view, and others where multiple streams get interleaved. Example:
![Screenshot_1577772788](https://user-images.githubusercontent.com/28173/71611928-b3f8e500-2b51-11ea-80c7-8153c4f37d04.png)

---

In fact, on such a message list it's possible for the stream name to get cut off too, so this solves a case of the stream analogue of #3631:
![Screenshot_1577773122](https://user-images.githubusercontent.com/28173/71612050-5add8100-2b52-11ea-8497-d3356c2829b5.png)

---

So I'm pretty sure on an interleaved message list we should include the stream name here.

Maybe on the other kind we can leave it out. I'm pretty confident we want *some* kind of "this is a topic" context, though, for the reasons mentioned in the earlier review:

> * The way this text ends up looking, it feels to me like it's likely to be confusing. It's sort of gray and without context.
> * Consulting some UI guidelines for these action sheets: [...]
>   
>   I see that the spot this is occupying is really meant for... basically asking the question that the options in the sheet are answers to.

In particular when the topic is short and generic, it can have pretty confusing alternate readings when the UI context just presents it as such a question. Scanning some real topics on czo -- and skipping the ones that are made *extra* confusing by the discussion being about developing Zulip -- there's "discussion", "new streams", "welcome", "Steve", among others.

Anyway, I think this UI looks good enough to merge!

gnprice commented:
> > * Check into how the new TS version interacts with our type-checking, and in particular see if there's anything we should do to our libdefs so we keep reasonable type-checking for our use of this library.
> 
> Objection: statement assumes facts not in evidence. We don't actually _have_ reasonable type-checking for this library as it stands – v1.1.0 has a `.flowconfig`, but doesn't seem to export any Flow types or provide any typing information. 

:+1: Thanks for the investigation -- that answers the question! I don't think I was making that assumption, though; evaluating the status quo is part of the question I'd wanted to answer.


> On the other hand, that does mean that the upgrade won't make things any worse on that front. I'm not inclined to say type definitions are needed in this PR.

Yep, agreed.

Thanks for the review, and thanks again @hashirsarwar for the PR and the revisions! Merging shortly.

gnprice commented:
And merged, as:
51d85946d deps: Upgrade @expo/react-native-action-sheet to 3.4.0
b103acaab messageActionSheet: Show stream and topic as title on action sheet.

(The PR will show on GitHub as closed rather than merged; it appears the PR was set to not let maintainers edit, and so I wasn't able to update the branch first to help GitHub draw the connection between this PR and the commits pushed.)
